File: ios/ElectrodeApiImpl/APIImpls/Navigation/ENNavigationAPIImpl.m
//
//  ENNavigationAPIImpl.m
//  ErnRunner
//
//  Created by Lianci Liu on 6/18/19.
//  Copyright © 2019 Walmart. All rights reserved.
//

#import "ENNavigationProtocol.h"

@implementation ENNavigationAPIImpl
+ (instancetype _Nonnull)sharedInstance {
    static dispatch_once_t onceToken;
    static id sharedInstance;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype) init {
    if (self = [super init]) {
        self.navigationAPI = [[EnNavigationAPI alloc] init];
        [self registerNavigationRequestHandler];
        [self registerBackRequestHandler];
        [self registerUpdateRequestHandler];
        [self registerFinishRequestHandler];
    }
    return self;
}

- (void)registerFinishRequestHandler {
    (void) [self.navigationAPI.requests registerFinishRequestHandlerWithHandler:^(id  _Nullable data, ElectrodeBridgeResponseCompletionHandler  _Nonnull block) {
        NSDictionary *routeData =  [data toDictionary];
        if (self.delegate && [self.delegate respondsToSelector:@selector(handleFinishFlow:withBlock:)]) {
            [self.delegate handleFinishFlow:routeData[@"finalPayload"] withBlock:^(NSString * _Nonnull messageCompletion) {
                block(messageCompletion, nil);
                return ;
            }];
        }
    }];
}

- (void)registerUpdateRequestHandler {
    (void) [self.navigationAPI.requests registerUpdateRequestHandlerWithHandler:^(id  _Nullable data, ElectrodeBridgeResponseCompletionHandler  _Nonnull block) {
        NSDictionary *routeData = [data toDictionary];
        if(self.delegate && [self.delegate respondsToSelector:@selector(updateNavigationBar:withBlock:)]) {
            NavigationBar *navBar = [[NavigationBar alloc] initWithDictionary:routeData[@"navigationBar"]];
            [self.delegate updateNavigationBar:navBar withBlock:^(NSString * _Nonnull message) {
                block(message, nil);
                return;
            }];
        }
        return;
    }];
}

- (void)registerBackRequestHandler {
    (void) [self.navigationAPI.requests registerBackRequestHandlerWithHandler:^(id  _Nullable data, ElectrodeBridgeResponseCompletionHandler  _Nonnull block) {
        NSDictionary *routeData = [data toDictionary];
        NSString *path = routeData == nil ? nil : routeData[@"path"];
        if(self.delegate && [self.delegate respondsToSelector:@selector(popToViewControllerWithPath:withBlock:)]) {
            [self.delegate popToViewControllerWithPath: path withBlock:^(NSString * _Nonnull message) {
                block(message, nil);
                return;
            }];
        }
        return;
    }];
}

- (void)registerNavigationRequestHandler {
    (void) [self.navigationAPI.requests registerNavigateRequestHandlerWithHandler:^(id  _Nullable data, ElectrodeBridgeResponseCompletionHandler  _Nonnull block) {
        NSDictionary *routeData =  [data toDictionary];
        NSString *path = routeData[@"path"];
        if ([path isEqualToString:@"finishFlow"]) {
            if (self.delegate && [self.delegate respondsToSelector:@selector(handleFinishFlow:withBlock:)]) {
                [self.delegate handleFinishFlow:routeData[@"jsonPayload"] withBlock:^(NSString * _Nonnull messageCompletion) {
                    block(messageCompletion, nil);
                    return ;
                }];
            }
        } else {
            if (self.delegate && [self.delegate respondsToSelector:@selector(handleNavigationRequestWithPath:withBlock:)]) {
                [self.delegate handleNavigationRequestWithPath:routeData withBlock:^(NSString * _Nonnull messageCompletion) {
                    block(messageCompletion, nil);
                    return;
                }];
            }
        }
    }];
}
@end

File: ios/ElectrodeApiImpl/APIImpls/Navigation/ENNavigationAPIImpl.swift
/*
 * Copyright 2019 Walmart Labs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

class ENNavigationAPIImpl: NSObject {
    static let shared = ENNavigationAPIImpl()
    let navigationAPI: EnNavigationAPI
    weak var currentViewController: ENNavigationProtocol?

    private override init() {
        self.navigationAPI = EnNavigationAPI()
        super.init()
        self.registerFinishRequestHandler()
        self.registerUpdateRequestHandler()
        self.registerBackRequestHandler()
        self.registerNavigationRequestHandler()
    }

    func registerFinishRequestHandler() {
        _ = self.navigationAPI.requests.registerFinishRequestHandler(handler: { (data, block) in
            let finishFlow = data as? String
            self.currentViewController?.handleFinishFlow(finalPayLoad: finishFlow, completion: { (messageCompletion) in
                block(messageCompletion, nil)
                return
            })
        })
    }

    func registerUpdateRequestHandler() {
        _ = self.navigationAPI.requests.registerUpdateRequestHandler(handler: { (data, block) in
            if let d = data as? ErnNavRoute, let ernData = d.toDictionary() as? [AnyHashable : Any] {
                if let navBarDict = ernData["navigationBar"] as? [AnyHashable : Any] {
                    let navBar = NavigationBar(dictionary: navBarDict)
                    self.currentViewController?.updateNavigationBar(navBar: navBar, completion: { (message) in
                        if message == "success" {
                            return block(message, nil)
                        } else {
                            let failureMessage = ElectrodeBridgeFailureMessage.createFailureMessage(withCode: "error", message: message)
                            return block(nil, failureMessage)
                        }
                    })
                }
            }
        })
    }

    func registerBackRequestHandler() {
        _ = self.navigationAPI.requests.registerBackRequestHandler(handler: { (data, block) in
            let d = data as? ErnNavRoute
            let ernData = d?.toDictionary() as? [AnyHashable : Any]
            self.currentViewController?.popToViewControllerWithPath(ernNavRoute: ernData, completion: { (message) in
                if message == "success" {
                    return block(message, nil)
                } else {
                    let failureMessage = ElectrodeBridgeFailureMessage.createFailureMessage(withCode: "error", message: message)
                    return block(nil, failureMessage)
                }
            })
        })
    }

    func registerNavigationRequestHandler() {
        _ = self.navigationAPI.requests.registerNavigateRequestHandler(handler: { (data, block) in
            if let d = data as? ErnNavRoute, let ernData = d.toDictionary() as? [AnyHashable : Any] {
                self.currentViewController?.handleNavigationRequestWithPath(routeData: ernData, completion: { (message) in
                    return block(message, nil)
                })
            }
        })
    }
}
